### PR TITLE
Pull Request for #108: Fix unit tests in the current builders

### PR DESCRIPTION
**Pull request by yuchuan2016**
I have fixed the unit tests failure. I have modified the test scripts for `build_r`, `build_stata`, `log` and `configuration_tests`. It would be great if @stanfordquan can take a loot at the stata and config tests, and @Shun-Yang can take a look at the r and log tests, as well as run these tests to see if they work. Thanks!

### CODE REVIEW

File: gslab_scons/tests/_side_effects.py [hunk 1]
```diff
@@ -34,8 +34,8 @@ def side_effect(*args, **kwargs):
             source = match.group('source')
             log    = '%s.Rout' % re.sub('\.R', '', source)
     
-        if executable == "R CMD BATCH" and log:
-            with open(log.strip(), 'wb') as log_file:
+        if executable == "Rscript" and log:
```
Comment (Shun-Yang) [inline gslab_scons/tests/_side_effects.py]: Can you explain a bit what this chunk is doing?

File: gslab_scons/tests/_test_helpers.py [hunk 2]
```diff
@@ -46,18 +46,22 @@ def command_match(command, executable, which = None):
                          command)
 
     elif executable in ['r', 'R']:
-        # e.g. "R CMD BATCH --no-save '--args cl_arg' script.R script.log"
+        # e.g. "Rscript --no-save --no-restore --verbose script.R > script.log 2>&1"
```
Comment (Shun-Yang) [inline gslab_scons/tests/_test_helpers.py]: `"Rscript --no-save --no-restore --verbose script.R args > script.log 2>&1"`

File: gslab_scons/tests/_test_helpers.py [hunk 3]
```diff
@@ -46,18 +46,22 @@ def command_match(command, executable, which = None):
...
+                         '\s*'
+                         '(?P<option3>--verbose)'
                          '\s*'
                          '(?P<source>[\.\/\w]+\.[rR])'
                          '\s*'
-                         '(?P<log>[\.\/\w]+(\.\w+)?)?',
-                         command)  
+                         '(?P<args>(\s?[\.\/\w]+)*)?'
+                         '\s*'
+                         '(?P<log>> [\.\/\w]+(\.\w+)?)?',
```
Comment (Shun-Yang) [inline gslab_scons/tests/_test_helpers.py]: I'm not familiar with regex. However, just wonder that does `'(?P<log>> [\.\/\w]+(\.\w+)?)?'` here match to `> script.log 2>&1` ?

File: gslab_scons/tests/_side_effects.py (hunk 1 above)
Comment (yuchuan2016) [inline gslab_scons/tests/_side_effects.py]: I'm not very familiar with `mock` either. But the basic idea is it does not actually run the command, but instead define what `side effect` this command will produce. Suppose we are running the following command from terminal:
`Rscript --no-save --no-restore --verbose script.R > script.log 2>&1`.

The effect of this command would be to run the script `script.R` and produce the log file `script.log`. When testing using `mock`, we do not actually run the command, but instead we define the behavior of this command in general (that is, to parse the command, get the executable, source, log, args, and do what the real command does). Then when tested, it works as if the command is run from the terminal. Does it make sense?

Comment (yuchuan2016) [inline gslab_scons/tests/_side_effects.py]: But I think I do need to change the above definition of log when log is missing.

File: gslab_scons/tests/_test_helpers.py (hunk 2 above)
Comment (yuchuan2016) [inline gslab_scons/tests/_test_helpers.py]: I think this is an example that would run in reality. Can you give me a real example of `args` to replace `args`?

File: gslab_scons/tests/_test_helpers.py (hunk 3 above)
Comment (yuchuan2016) [inline gslab_scons/tests/_test_helpers.py]: No. If the command is ```Rscript --no-save --no-restore --verbose script.R > script.log 2>&1```, the output would be

```
match.group('executable') = "Rscript"
...
match.group('log') = "> script.log"
```
The parser just want to get the file name of the log file.

File: gslab_scons/tests/_test_helpers.py (hunk 2 above)
Comment (Shun-Yang) [inline gslab_scons/tests/_test_helpers.py]: What about `"Rscript --no-save --no-restore --verbose script.R input.txt > script.log 2>&1"`?

Comment (yuchuan2016) [inline gslab_scons/tests/_test_helpers.py]: Thanks! Changed.

File: gslab_scons/tests/_test_helpers.py (hunk 3 above)
Comment (yuchuan2016) [inline gslab_scons/tests/_test_helpers.py]: Changed to include `2>&1` and add check [here](https://github.com/gslab-econ/gslab_python/pull/109/files#diff-53ecb6f88f15102dbd104d4d960bda14R38).

**Comment by yuchuan2016** [comment]
@Shun-Yang , I have fixed most test failures brought in by #110. However, I find it hard to fix the tests for `build_tables`. Currently I comment out the `check_target` line in `build_tables`. We could perhaps revisit this later.

**Comment by stanfordquan** [comment]
Looks good @yuchuan2016 , except for the test you mentioned. 

This is failing on Windows. In fact, I couldn't get the `python setup.py test` syntax working at all and the `python -m unittest discover` syntax from within the test directory is returning all kinds of error. I vote we revisit at the end.

File: gslab_scons/tests/test_build_stata.py [hunk 4]
```diff
@@ -47,32 +47,14 @@ def test_windows(self, mock_is_64, mock_check, mock_path):
...
         when given appropriate inputs. 
         '''
-        # i) Non-64-bit Windows without a STATAEXE environment variable
-        mock_check.side_effect = fx.make_stata_side_effect('statamp.exe')
-        mock_path.side_effect  = fx.make_stata_path_effect('statamp.exe')
+        mock_check.side_effect = fx.make_stata_side_effect('StataMP-64.exe')
+        mock_path.side_effect  = fx.make_stata_path_effect('StataMP-64.exe')
         mock_is_64.return_value = False
 
         env = {'stata_executable' : None}
```
Comment (stanfordquan) [inline gslab_scons/tests/test_build_stata.py]: may want to try this for the current set of values that trigger default options being `None, 'None', ''`

**Comment by yuchuan2016** [comment]
@stanfordquan , thanks! `python setup.py test` just could not work on my PC either. I agree that we want to put this at the end. Do you think we should merge this branch back to 4.1.0 or we wait until the end?

**Comment by stanfordquan** [comment]
I think this is granular enough that we can pull it in.

**Comment by yuchuan2016** [comment]
Merge back for now. Will update unit tests in #112

**Pull request merged**